Generative AI 2025 chapter (#4301)
HTTPArchive/almanac.httparchive.org: src/content/en/2025/generative-ai.md

* Graphics Processing Unit (GPU): High throughput, ideal for AI-accelerated digital content creation.
* Neural Processing Unit (NPU) or Tensor Processing Unit (TPU): Low power, ideal for sustained AI workloads and AI offload for battery life.

The three key APIs that facilitate local AI inference are WebAssembly (CPU), WebGPU (GPU), and WebNN (CPU, GPU, and NPU).

All statistics in this section refer to the percentage of Google Chrome page loads that utilize these APIs at least once, across all release channels and platforms. It is essential to note that the use of WebAssembly and WebGPU does not confirm AI activity; these are general-purpose APIs frequently utilized for tasks such as complex calculations, 3D visualizations, or gaming.

##### WebAssembly

[WebAssembly](https://developer.mozilla.org/en-US/docs/WebAssembly) (Wasm) acts as the bytecode for the web. Code written in various programming languages, including C++ and Rust, can be compiled into WebAssembly. It enables developers to write optimized, high-performance code that is executed by the browser's scripting engine on the user's CPU.

<!-- markdownlint-disable-next-line MD051 -->

> WebAssembly has broad browser support, being implemented in all relevant browser engines since 2017. In 2025, the usage of WebAssembly (i.e., the instantiation of a WebAssembly module relative to page loads) experienced 27% linear growth (see [Figure 1](#fig-1)) from being active on 4.44% of page loads in January, and 5.64% in December 2025 (January 2024: 3.37%).

Reviewer: Technically it's figure 7.1, but you might be better saying something like "see the following figure" as much less likely to get out of sync if you add or remove figures.

```suggestion
WebAssembly has broad browser support, being implemented in all relevant browser engines since 2017. In 2025, the usage of WebAssembly (i.e., the instantiation of a WebAssembly module relative to page loads) experienced 27% linear growth (see [Figure 7.1](#fig-1)) from being active on 4.44% of page loads in January, and 5.64% in December 2025 (January 2024: 3.37%).
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTXSz19p32lprObXmLblQxEy5u0Sjd9QajNixDEOJutiaxi1aWk37ojoY5Z3D-GfgHg3Ggu23VZS2yI/pubchart?oid=148205691&format=interactive",
  sheets_gid="303424003"
  )
}}

##### WebGPU

[WebGPU](https://developer.mozilla.org/en-US/docs/Web/API/WebGPU_API) is the modern successor to [WebGL](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API), designed to expose the capabilities of modern GPUs to the web. Unlike WebGL, which was strictly for graphics, WebGPU provides support for "compute shaders", allowing for general-purpose computing on graphics cards. This allows developers to perform massive parallel calculations, as required by AI models, directly on the user's graphics card.

<!-- markdownlint-disable-next-line MD051 -->
WebGPU has become the standard foundation for running AI workloads in the browser. With the release of Firefox 141 in November 2025, [WebGPU became available in all relevant browser engines](https://web.dev/blog/webgpu-supported-major-browsers) (Chromium, Gecko, and WebKit). In 2025, WebGPU experienced exponential growth in page loads, increasing by 147% (see [Figure 2](#fig-2)).

> The July 2025 crawl of HTTP Archive data suggested that the API is used on 55,825 desktop sites and 68,043 mobile sites, a significant increase from 4,479 desktop and 4,742 mobile pages from the July 2024 crawl.

Reviewer: Better to talk in percentages than absolute numbers:
https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Analysts'-Guide#a-warning-on-totals

{{ figure_markup(
  image="genai-webllm-downloads.png",
  caption="npm package downloads of @mlc-ai/web-llm.",
  description="Bar chart showing npm downloads for the @mlc-ai/web-llm package from January to November 2025. Downloads increased by 340% from 23,425 package downloads in January to 103,084 downloads in November, with a significant spike occurring in August where downloads nearly doubled from 57,114 in July to 102,440 in August.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTXSz19p32lprObXmLblQxEy5u0Sjd9QajNixDEOJutiaxi1aWk37ojoY5Z3D-GfgHg3Ggu23VZS2yI/pubchart?oid=672596885&format=interactive",
  sheets_gid="87936740"
  )
}}

<a hreflang="en" href="https://huggingface.co/docs/transformers.js/index">Transformers.js</a>, developed by <a hreflang="en" href="https://huggingface.co/">Hugging Face</a>, functions as a comprehensive JavaScript library that mirrors the popular Python `transformers` API. Under the hood, it relies on ONNX Runtime Web to execute. It allows developers to run pre-trained models for various tasks through simple high-level pipelines, not just LLMs.

<!-- markdownlint-disable-next-line MD051 -->

> From January to November 2025, the npm package downloads almost tripled, also with a notable surge in August 2025 (see [Figure 6](#fig-6)).

Reviewer: ```suggestion
From January to November 2025, the npm package downloads almost tripled, also with a notable surge in August 2025 (see [Figure 7.6](#fig-6)).
```

[Built-in AI](https://developer.chrome.com/docs/ai/built-in) is an initiative by Google and Microsoft that aims to provide high-level AI APIs to web developers. Unlike BYOAI, this approach leverages models facilitated by the browser itself. While developers cannot specify an exact model, this method allows all websites to share the same model, meaning it only needs to be downloaded once.

The initiative consists of multiple APIs:

* Prompt API: Gives developers low-level access to a local LLM.
* Task-specific APIs:
  * Writing Assistance APIs: Summarizing, writing, and rewriting text.
  * Proofreader API: Finding and fixing mistakes in text.
  * Language Detector and Translator APIs: Detecting the language of text content and translating it into another language.

All APIs are specified within the WebML Community Group, meaning they are still incubating and not yet on the W3C Recommendations track. Some of the APIs are already generally available, while others are in Origin Trial or need a browser flag to be enabled—unlike WebGPU, which is now stable and widely available for shipping production features.

<!-- markdownlint-disable-next-line MD051 -->

> The package of TypeScript typings for Built-in AI (see [Figure 7](#fig-7)) may reflect the experimental status of the APIs. Downloads peaked in August 2025 with 25,770 downloads, followed by a gradual decline. This trend may suggest that developers are hesitant to adopt experimental APIs. However, the download statistics for a typings package may not reflect the actual usage of the API.

Reviewer: ```suggestion
The package of TypeScript typings for Built-in AI (see [Figure 7.7](#fig-7)) may reflect the experimental status of the APIs. Downloads peaked in August 2025 with 25,770 downloads, followed by a gradual decline. This trend may suggest that developers are hesitant to adopt experimental APIs. However, the download statistics for a typings package may not reflect the actual usage of the API.
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTXSz19p32lprObXmLblQxEy5u0Sjd9QajNixDEOJutiaxi1aWk37ojoY5Z3D-GfgHg3Ggu23VZS2yI/pubchart?oid=1897808878&format=interactive",
  width="621",
  height="845",
  sheets_gid="342591090",
  sql_file="user-agent-named.sql"
  )
}}

The other popular AI bots are the ones by Anthropic (`claudebot`, `anthropic-ai`, `claude-web`), Google (`google-extended`), OpenAI (`chatgpt-user`), and Perplexity (`perplexitybot`), among other less popular ones. The prevalence of AI bots is much higher on `robots.txt` of popular websites, almost all of them disallowing the bots. This is likely because popular websites can afford to block AI crawlers, without significantly hurting their website traffic due to their popularity. Overall, we observed an increase in adoption of AI User-Agent bots in `robots.txt` files.

> ### llms.txt

Reviewer: ```suggestion
### `llms.txt`
```

<!-- markdownlint-disable-next-line MD051 -->
Vercel dominated, though it offered hosting under subdomains long before the release of its vibe coding tool v0 in late 2023, and the data showed no noticeable surge afterward. Lovable, on the other hand, grew from only 10 subdomains in our data at the beginning of 2025 to 315 by October 2025 (see [Figure 16](#fig-16)).

{{ figure_markup(
  image="genai-vibe-coding-subdomains.png",
  caption="Subdomains of vibe coding platforms.",
  description="Line chart comparing the count of vibe coding platform subdomains relative to the entire HTTP Archive dataset, based on quarterly crawls since Q1 2020. Vercel leads the trend, rising steadily from 2021 to reach nearly 0.04% of all sites by October 2025, while newer entrants like Lovable and Replit show visible adoption beginning in early 2025.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTXSz19p32lprObXmLblQxEy5u0Sjd9QajNixDEOJutiaxi1aWk37ojoY5Z3D-GfgHg3Ggu23VZS2yI/pubchart?oid=1670163713&format=interactive",
  width="797",
  height="417",
  sheets_gid="2051033428",
  sql_file="vibecodetools.sql"
  )
}}

> ### The .ai domain

Reviewer: ```suggestion
### The `.ai` domain
```

<!-- markdownlint-disable-next-line MD051 -->
While both AI and the `.ai` domain existed long before ChatGPT, the new possibilities that emerged in its wake led to a wave of AI-native businesses. Many of them adopted the `.ai` top-level domain, the country code for Anguilla. This resulted in a substantial rise in `.ai` domain records across all ranks, as shown in [Figure 17](#fig-17) below comparing the pre-ChatGPT moment in 2022 with usage in 2025.

> The only `.ai` domains that made it into the top 1k most visited websites in the Chrome dataset were https://character.ai and a single adult site.

Reviewer: ```suggestion
The only `.ai` domains that made it into the top 1,000 most visited websites in the Chrome dataset were https://character.ai and a single adult site.
```

  classes="big-number",
) }}

The <a hreflang="en" href="https://webmachinelearning.github.io/">Web Machine Learning</a> Community Group and Working Group of the <a hreflang="en" href="https://www.w3.org/">World Wide Web Consortium</a> (W3C) are actively standardizing this shift to make AI a "first-class web citizen." This effort follows two primary architectural directions: *Bring Your Own AI* and *Built-in AI*.

#### Bring Your Own AI

In the Bring Your Own AI (BYOAI) approach, the developer is responsible for shipping the model to the user. The web application downloads a specific model binary and executes it using low-level APIs on local hardware. This allows running highly specialized models, tailored to the use case, but also requires significant bandwidth.

There are <a hreflang="en" href="https://www.w3.org/2024/01/webevolve-series-events/media/slides/hu-ningxin.pdf#page=4">three processing units</a> that can be used to run AI inference locally:

* *Central Processing Unit* (CPU): Responds quickly, ideal for low-latency AI workloads.
* *Graphics Processing Unit* (GPU): High throughput, ideal for AI-accelerated digital content creation.

> * *Neural Processing Unit* (NPU) or *Tensor Processing Unit* (TPU): Low power, ideal for sustained AI workloads and AI offload for battery life.

Reviewer: ```suggestion
- _Central Processing Unit_ (CPU): Responds quickly, ideal for low-latency AI workloads.
- _Graphics Processing Unit_ (GPU): High throughput, ideal for AI-accelerated digital content creation.
- _Neural Processing Unit_ (NPU) or _Tensor Processing Unit_ (TPU): Low power, ideal for sustained AI workloads and AI offload for battery life.
```

The three key APIs that facilitate local AI inference are WebAssembly (CPU), WebGPU (GPU), and WebNN (CPU, GPU, and NPU).

It is essential to note that the use of WebAssembly and WebGPU does not confirm AI activity; these are general-purpose APIs frequently utilized for tasks such as complex calculations, 3D visualizations, or gaming.

##### WebAssembly

> [WebAssembly](https://developer.mozilla.org/en-US/docs/WebAssembly) acts as the bytecode for the web. Code written in various programming languages, including C++ and Rust, can be compiled into WebAssembly. It enables developers to write optimized, high-performance code that is executed by the browser's scripting engine on the user's CPU.

Reviewer: We exclude the `/en-US` from MDN links so it can redirect to the native language version if it exists (which, FYI, is also why we don't include the `hreflang=en` attribute unlike other links, and so can use markdown links like you did here):

```suggestion
[WebAssembly](https://developer.mozilla.org/docs/WebAssembly) acts as the bytecode for the web. Code written in various programming languages, including C++ and Rust, can be compiled into WebAssembly. It enables developers to write optimized, high-performance code that is executed by the browser's scripting engine on the user's CPU.
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTXSz19p32lprObXmLblQxEy5u0Sjd9QajNixDEOJutiaxi1aWk37ojoY5Z3D-GfgHg3Ggu23VZS2yI/pubchart?oid=148205691&format=interactive",
  sheets_gid="303424003"
  )
}}

##### WebGPU

[WebGPU](https://developer.mozilla.org/en-US/docs/Web/API/WebGPU_API) is the modern successor to [WebGL](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API), designed to expose the capabilities of modern GPUs to the web. Unlike WebGL, which was strictly for graphics, WebGPU provides support for *compute shaders*, allowing for general-purpose computing on graphics cards. This allows developers to perform massive parallel calculations, as required by AI models, directly on the user's graphics card.

WebGPU has become the standard foundation for running AI workloads in the browser. With the release of Firefox 141 in November 2025, [WebGPU became available in all relevant browser engines](https://web.dev/blog/webgpu-supported-major-browsers) (Chromium, Gecko, and WebKit).

> The July 2025 crawl of HTTP Archive data shows that the API is used on 0.243% of all desktop sites and 0.238% of mobile sites. This represents a significant increase of 591% (up from 0.035%) on desktop and 709% (up from 0.029%) on mobile compared to the July 2024 crawl (see figure below). <a hreflang="en" href="https://chromestatus.com/metrics/feature/timeline/popularity/3888">Chrome Platform Status data</a> suggests an exponential growth in activations per page load, increasing by 147% over the course of 2025.

Reviewer: ```suggestion
The July 2025 crawl of HTTP Archive data shows that the API is used on 0.243% of all desktop sites and 0.238% of mobile sites. While still small overall, this represents a significant increase of 591% (up from 0.035%) on desktop and 709% (up from 0.029%) on mobile compared to the July 2024 crawl (see figure below). <a hreflang="en" href="https://chromestatus.com/metrics/feature/timeline/popularity/3888">Chrome Platform Status data</a> suggests an exponential growth in activations per page load, increasing by 147% over the course of 2025.
```

### `robots.txt`

The <a hreflang="en" href="https://www.robotstxt.org/">`robots.txt` file</a>, located at the root of a domain (for example, `http://example.com/robots.txt`), allows website owners to declare crawl directives for automated bots. Historically, web crawling was primarily performed by search engines and archives. However, in the age of Generative AI, websites can also be crawled by AI agents, or by model providers collecting internet-scale data to train their LLMs. Consequently, websites increasingly rely on `robots.txt` files to manage access for these crawlers.

An example directive targeting the user agent `GPTBot`, <a hreflang="en" href="https://platform.openai.com/docs/bots">used by OpenAI for model training</a>, is shown below:

```
User-Agent: GPTBot
Disallow: /
```

It is important to note that adherence to `robots.txt` is voluntary and does not strictly enforce access control.

Usage of `robots.txt` remains highly prevalent: approximately 94.1% of the \~12.9 million sites analyzed include a `robots.txt` file with at least one directive (see the following figure).

{{ figure_markup(

>   image="genai-robots-txt-adoption.png",

Reviewer: Tweak the chart (made the chart area smaller, and lowered to 1 decimal place). Please retake image.

  caption="`robots.txt` adoption.",

>   description="Bar chart comparing sites with and without `robots.txt`. A significant 94.12% of sites serve the file.",

Reviewer: ```suggestion
  description="Bar chart comparing sites with and without `robots.txt`. A significant 94.1% of sites serve the file.",
```

### AI colors

Models can tend to overrepresent certain words, which then serve as an AI fingerprint. Researchers at Florida State University compared research papers published on PubMed between 2020 and 2024. <a hreflang="en" href="https://aclanthology.org/2025.coling-main.426.pdf">Their analysis</a> revealed a staggering 6,697% increase in the use of the word "delves." The other conjugations of the verb "delve" also increased significantly. While those kinds of indicators should never be used to classify individual cases as AI-generated, they are still useful for understanding the trend of adoption. In this part of the chapter, we will delve into ;) common patterns used by LLMs to generate websites.

{{ figure_markup(
  content="6,697%",
  caption="Usage increase of the word \"delves\" in research papers published on PubMed.",
  classes="big-number",
) }}

The most recognizable indicator of AI-generated web design is the pervasive use of purple and gradients, often referred to as "the AI <a hreflang="en" href="https://ai-engineering-trend.medium.com/the-mystery-behind-ais-purple-problem-revealed-0234afdb292e">Purple Problem</a>."

> Adam Wathan, creator of the widely adopted CSS framework <a hreflang="en" href="https://tailwindcss.com/">Tailwind</a>, noted <a hreflang="en" href="https://youtu.be/AG_791Y-vs4?t=86">in a recent interview</a> that this trend likely stems from the timing of major AI training crawls. In the early 2020s, the Tailwind team used deep purples extensively in their documentation and examples to reflect contemporary design trends. Consequently, LLMs trained on that data seem to have "locked in" an aesthetic that has since been replaced by more modern trends, such as the current preference for black designs.

Reviewer: ```suggestion
[Adam Wathan](https://x.com/adamwathan), creator of the widely adopted CSS framework <a hreflang="en" href="https://tailwindcss.com/">Tailwind</a>, noted <a hreflang="en" href="https://youtu.be/AG_791Y-vs4?t=86">in a recent interview</a> that this trend likely stems from the timing of major AI training crawls. In the early 2020s, the Tailwind team used deep purples extensively in their documentation and examples to reflect contemporary design trends. Consequently, LLMs trained on that data seem to have "locked in" an aesthetic that has since been replaced by more modern trends, such as the current preference for black designs.
```

Author: No `<a hreflang="en">`?

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTXSz19p32lprObXmLblQxEy5u0Sjd9QajNixDEOJutiaxi1aWk37ojoY5Z3D-GfgHg3Ggu23VZS2yI/pubchart?oid=1045089879&format=interactive",
  width="642",
  height="417",
  sheets_gid="1065437374"
  )
}}

However, due to the explosive overall growth of the Tailwind framework, these specific colors saw a significant rise when measured as a percentage of the entire web, as shown by the following figure.

{{ figure_markup(
  image="genai-ai-colors.png",
  caption="Usage of \"AI colors\" over the years.",
  description="Line chart comparing the use of the aforementioned AI colors and indigo-500 across all pages, based on quarterly HTTP Archive crawls since Q3 2022. Their use rose from 0.02% (purple) and 0.03% (indigo-500) in Q3 2022 to 0.49% (purple) and 0.54% (indigo-500) in Q4 2025, with indigo-500 becoming slightly more popular.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTXSz19p32lprObXmLblQxEy5u0Sjd9QajNixDEOJutiaxi1aWk37ojoY5Z3D-GfgHg3Ggu23VZS2yI/pubchart?oid=428756366&format=interactive",
  width="642",
  height="417",

>   sheets_gid="1065437374",

Reviewer: Ditto no sql file
